protostar/cheatable_starknet/cheaters/expect_events_controller.py

```diff
@@ -0,0 +1,166 @@
+    from_address: Address
+    key: Selector
+    data: Optional[CairoData] = None
+
+
+@dataclass
+class EventMatching:
+    pass
+
+
+@dataclass
+class SkippedEventMatching(EventMatching):
+    emitted_event: Event
+
+
+@dataclass
+class AcceptedEventMatching(EventMatching):
+    emitted_event: Event
+    expected_event: Event
+
+
+@dataclass
+class FailedEventMatching(EventMatching):
+    expected_event: Event
+
+
+@dataclass
+class EventMatchingResult:
+    event_matchings: list[EventMatching]
+
+    @property
+    def failed_event_matchings(self):
+        return [
+            matching
+            for matching in self.event_matchings
+            if isinstance(matching, FailedEventMatching)
+        ]
+
+    @property
+    def should_be_accepted(self) -> bool:
+        return len(self.failed_event_matchings) == 0
+
+
+class ExpectEventsController:
+    def __init__(
+        self,
+        test_finish_hook: "Hook",
+        test_execution_state: "CairoTestExecutionState",
+        cheatable_starknet_facade: "CheatableStarknetFacade",
+    ) -> None:
+        self._test_execution_state = test_execution_state
+        self._cheatable_starknet_facade = cheatable_starknet_facade
+        self._test_finish_hook = test_finish_hook
+        self._test_finish_hook.on(self.compare_expected_and_actual_results)
+
+    def add_expected_events(self, expected_events: list[Event]):
+        self._test_execution_state.expected_events_list.append(expected_events)
+
+    def compare_expected_and_actual_results(self):
+        for expected_events in self._test_execution_state.expected_events_list:
+            matching_result = match_events(
+                expected_events=expected_events,
+                emitted_events=self._cheatable_starknet_facade.get_emitted_events(),
+            )
+            if not matching_result.should_be_accepted:
+                raise ExpectEventsMismatchReportedException(matching_result)
+
+
+def match_events(
+    expected_events: list[Event], emitted_events: list[Event]
+) -> EventMatchingResult:
+    event_matchings: list[EventMatching] = []
+    for emitted_event in emitted_events:
+        if len(expected_events) == 0 or not should_accept_event_matching(
+            expected_event=expected_events[0], emitted_event=emitted_event
+        ):
+            event_matchings.append(SkippedEventMatching(emitted_event=emitted_event))
+        else:
+            event_matchings.append(
+                AcceptedEventMatching(
+                    emitted_event=emitted_event,
+                    expected_event=expected_events[0],
+                )
+            )
+            expected_events.pop(0)
+    failed_event_matchings = [
+        FailedEventMatching(expected_event) for expected_event in list(expected_events)
+    ]
+    return EventMatchingResult(
+        event_matchings=[*event_matchings, *failed_event_matchings],
+    )
+
+
+def should_accept_event_matching(expected_event: Event, emitted_event: Event) -> bool:
```

Comment:
`events_match` would be a nicer name

Reply:
Or even better, we could add a `.matches()` method to the `Event` dataclass

Reply:
I'd avoid adding it to the data class because it's not obvious whether expected_event should match emitted_event or emitted_event should match expected_event.

Reply:
I'd keep this name for the consistency reasons — there's also`EventMatching` data structure.